File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章灵能传输.cpp
//灵能传输  lanqiaoOJ 196 

#include<bits/stdc++.h>
using namespace std;
const int N = 3e5;
long long a[N],s[N];
bool vis[N];
int main(){
    int T;    scanf("%d",&T);
    while(T--){
        memset(vis,0,sizeof(vis));
        int n;   scanf("%d",&n);
        s[0]=0;
        for(int i=1;i<=n;++i){scanf("%lld",&s[i]);  s[i] += s[i-1]; } //计算前缀和
        long long s0=0,sn=s[n];
        if(s0>sn) swap(s0,sn);
        sort(s,s+n+1);
        int L=0,R=n;
        for(int i = lower_bound(s,s+n+1,s0) - s;i>=0;i-=2)
           a[L++]=s[i], vis[i]=1;     //图中的路线1: 从s0到min。隔一个数取一个

        for(int i = lower_bound(s,s+n+1,sn)-s;i<=n;i+=2)
            a[R--]=s[i], vis[i]=1;   //图中的路线3:  从max到sn。隔一个数取一个
        for(int i=0;i<=n;++i)        //图中的路线2：从min到max
            if(!vis[i])  a[L++]=s[i];
        long long res=0;
        for(int i=1;i<=n;++i)  res = max(res,abs(a[i]-a[i-1]));
        printf("%lld\n",res);
    }
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章灵能传输.py
#灵能传输  lanqiaoOJ 196

T=int(input())
for t in range(T):
    n=int(input())
    a=list(map(int,input().split()))
    s=[0]+a
    for i in range(1,n+1):  s[i] += s[i-1]   #前缀和
    s0=0
    sn = s[n]
    if s0>sn:  sn,s0 = s0,sn       #交换: swap(s0, sn)
    s.sort()        
    for i in range(n+1):           #找s[0]和s[n]的位置
        if s[i]==s0: s0 = i;  break
    for i in range(n,-1,-1):
        if s[i]==sn: sn = i;  break
    L, R = 0,n
    a=[0 for i in range(n+1)]
    a[n] = s[n]
    vis = [True for i in range(n+1)]
    for i in range(s0,-1,-2): a[L]=s[i]; L+=1; vis[i]=False
    for i in range(sn,n+1,2): a[R]=s[i]; R-=1; vis[i]=False        
    for i in range(n+1):
        if vis[i]: a[L]=s[i]; L+=1
    res = 0
    for i in range(n): res=max(res,abs(a[i+1]-a[i]))
    print(res)
